File: js-demo/mwua.js
function uniform(min, max) {
  return Math.random() * (max - min) + min;
}

function initializeWeights(n) {
  var weights = new Array(n);
  for (var i = 0; i < weights.length; i++) {
    weights[i] = 1;
  }

  return weights;
}

function draw(weights) {
  // Sample an index proportionally to the size of the number at that index
  var sum = 0;
  for (var i = 0; i < weights.length; i++) {
    sum += weights[i];
  }

  choice = uniform(0, sum(weights))
  choiceIndex = 0

  for (var i = 0; i < weights.length; i++) {
    weight = weights[i];
    choice -= weight;
    if (choice <= 0) {
        return choiceIndex;
    }

    choiceIndex += 1;
  }
}

function mwua(objects, observeOutcome, reward, learningRate, numRounds) {
  var weights = initializeWeights(objects.length);
  
  for (var i = 0; i < numRounds; i++) {
    chosenObjectIndex = draw(weights);
    chosenObject = objects[chosenObjectIndex];

    outcome = observeOutcome(t, chosenObject);
    thisRoundReward = reward(chosenObject, outcome);

    for (var j = 0; j < weights.length; j++) {
      weights[j] *= (1 + learningRate * reward(objects[j], outcome))
    }
  }

  return weights;
}

File: mwua.py
import random


# draw: [float] -> int
# pick an index from the given list of floats proportionally
# to the size of the entry (i.e. normalize to a probability
# distribution and draw according to the probabilities).
def draw(weights):
    choice = random.uniform(0, sum(weights))
    choiceIndex = 0

    for weight in weights:
        choice -= weight
        if choice <= 0:
            return choiceIndex

        choiceIndex += 1


# MWUA: the multiplicative weights update algorithm
def MWUA(objects, observeOutcome, reward, learningRate, numRounds):
    weights = [1] * len(objects)
    cumulativeReward = 0

    for t in range(numRounds):
        chosenObjectIndex = draw(weights)
        chosenObject = objects[chosenObjectIndex]

        outcome = observeOutcome(t, weights, chosenObject)
        thisRoundReward = reward(chosenObject, outcome)
        cumulativeReward += thisRoundReward

        for i in range(len(weights)):
            weights[i] *= (1 + learningRate * reward(objects[i], outcome))

    return weights
